shared/domains/chelonia/chelonia.js

```diff
@@ -1642,50 +1646,116 @@ export default (sbp('sbp/selectors/register', {
+          innerSigningKeyId,
+          encryptionKeyId,
+          signingKeyId,
+          data,
+          meta: key
+        })
+        response = await this.config.fetch(url, {
+          headers: new Headers([[
+            'authorization', buildShelterAuthorizationHeader.call(this, contractID)
+          ], [
+            'if-match', ifMatch || '""'
+          ]
+          ]),
+          method: 'POST',
+          body: JSON.stringify(serializedData),
+          signal: this.abortController.signal
+        })
+      } else {
+        // If no initial data provided, perform a GET `fetch` to get the current
+        // data and CID
+        response = await this.config.fetch(url, {
+          headers: new Headers([[
+            'authorization', buildShelterAuthorizationHeader.call(this, contractID)
+          ]]),
+          signal: this.abortController.signal
+        })
+      }
+      // The `resolveData` function is tasked with computing merged data, as in
+      // merging the existing stored values (after a conflict or initial fetch)
+      // and new data. The return value indicates whether there should be a new
+      // attempt at storing updated data (if `true`) or not (if `false`)
+      const resolveData = async () => {
+        let currentValue
+        if (response.ok || response.status === 409 || response.status === 412) {
+          const serializedData = await response.json()
+          currentValue = parseEncryptedOrUnencryptedMessage.call(this, {
+            contractID,
+            serializedData,
+            meta: key
+          })
+        } else if (response.status !== 404 && response.status !== 410) {
+          throw new ChelErrorUnexpectedHttpResponseCode('[kv/set] Invalid response code: ' + response.status)
+        }
+        const result = await onconflict({
+          contractID,
+          key,
+          failedData: data,
+          status: response.status,
+          etag: response.headers.get('x-cid') || response.headers.get('etag'),
+          get currentData () {
+            return currentValue?.data
+          },
+          currentValue
+        })
+        if (!result) return false
+
+        data = result[0]
+        ifMatch = result[1]
+        return true
+      }
+      if (!data && typeof onconflict === 'function') {
+        if (await resolveData()) {
+          continue
+        } else {
+          break
+        }
+      }
       if (!response.ok) {
-        if (response.status === 409) {
+        if (response.status === 409 || response.status === 412) {
           if (--maxAttempts <= 0) {
             throw new Error('kv/set conflict setting KV value')
           }
+          // Only retry if an onconflict handler exists to potentially resolve it
           await delay(randomIntFromRange(0, 1500))
           if (typeof onconflict === 'function') {
-            data = await onconflict(contractID, key, data)
-            if (!data) { break }
+            if (await resolveData()) {
```

Comment:
The check for `typeof onconflict === 'function'` (or the inverse) is done 3 times... maybe this condition could be cached and DRY'd?